### PR TITLE
fix: working TList serialization

### PR DESCRIPTION
**Pull request by lobis**
Closes https://github.com/scikit-hep/uproot5/issues/762.

This PR was motivated because of problems serializing `TList`. 

- `TList` serialization was fixed in 4da653fee0f28789fbb20abcad3a4541acd6701a.

A few other minor improvements were done along the way:

- `TObject` should now correctly serialize `fUniqueID` member (previously it was using only 0).
- `TObject` `__repr__` is fixed.
- A method `Model_TObject.empty` has been added as override to avoid using its parent method in order to simplify code.
- `TObjString` now correctly initializes its `bases` with a `TObject` with all its members.
- Updated some `tojson` calls of above mentined classes.
- Added / Updated tests for writing and reading `TObjString` and `TList`.

### CODE REVIEW

**Comment by jpivarski** [comment]
Since I saw your message in Slack, I started looking into this, but it seems that you're ahead of me: yes, the TObject will need its parameters. (Doesn't the TObject constructor make them, though?)

TObjString is heavily tested in the testing suite, but not TList. I'm a little surprised that TList is even implemented, but I suppose it might be because it was needed to implement something else, such as histograms. (Writing TObjString, descendants of TH1, and simple and jagged TTrees have been implemented; nothing else has been attempted.)

It's very likely that TList's `_serialize` method was only written for _empty_ TLists. It's very likely that some histogram class can hold a list of objects that we don't attempt to write (such as fit functions) and the TList serialization was only written to allow empty TLists to be included in histograms. So when you tried to write a non-empty TList, you were in uncharted territory.

Even though you've found something that works, you should remain suspicious of it because it might be working by accident. You're filling the `fUniqueID` and `fBits` of the TObject with zeros; the reader trying to deserialize a different part of the structure when it gets to those zeros, and interprets the zeros as that other thing, like size of the list, length of the TList `fName`, or something. Be sure to use much more than one test; lists of various lengths, names, etc.

What I was looking at before seeing this PR is the fact that each item in a TList has a `fOption`. The writer does a Python `zip` through items and `self._options`:

https://github.com/scikit-hep/uproot5/blob/baa82c94eba0e7155cbacfa3728ec121037f406f/src/uproot/models/TList.py#L103-L105

Python `zip` stops when it reaches the end of _either_ of the lists it is given, so filling `self._data` and not `self._options` will not get it to write the correct number of items. If that number of items disagrees with the declared `self._members["fSize"]`, the reader will likely crash.

In your testing, you can test read-back first with Uproot, then (when that works) with ROOT. Uproot is generally more forgiving of wrong formats, since it reads as little as it needs to and doesn't do all of the consistency checks that ROOT does. Also, it's easier to use Uproot in a debugger or insert print-outs (my favorite method of debugging, since Python loads quickly and doesn't have a long compilation phase), so you can more easily examine which bytes the reader is seeing when it's expecting to be building a given structure. Once it works in Uproot, you still have to test it in ROOT, of course, because it might still be wrong.

I hope this is useful! I'll keep tabs on your progress, and go ahead and ask questions on #awkward-uproot in Slack as you go. Good luck!

**Comment by agoose77** [comment]
@jpivarski this might be a good place to discuss; should a bare `TString` be round-trippable in Awkward? I noticed that it didn't serialise when taking a quick look at *this* issue with `TObjString`.

**Comment by lobis** [comment]
I also noticed this interesting behaviour that may explain why this has slipped under the radar:

```
import uproot

filename = "list.root"
key = "listOfStrings"

s = uproot.writing.identify.to_TObjString("test")
with uproot.recreate(filename) as f:
    try:
        print(s.tojson())  # this does not work
    except AttributeError as e:
        print(e)
    f[key] = s
    print(f[key].tojson())  # this works
```

**Comment by jpivarski** [comment]
> should a bare `TString` be round-trippable in Awkward?

I don't think that bare TString objects can be put in TDirectories (even in ROOT, that is). I think that's why TObjString exists: it inherits from TObject so that it has the `fProcess`, `fBits` infrastructure that objects-in-TDirectories need. (They might also need `fName` and `fTitle`, since this information is also in the TKeys of a TDirectory.) But perhaps Uproot could use some code to _prohibit_ non-TDirectory-ready objects from being written to TDirectories.

> I also noticed this interesting behaviour that may explain why this has slipped under the radar:

The `tojson` method is not actually used in serialization; JSON serialization is an alternative to binary serialization, defined by ROOT, and so I used it in the early part of the project to verify that I was at least identifying the right data to serialize, before attempting the binary format. Since those early tests ([test_0010-start-streamers.py](https://github.com/scikit-hep/uproot5/blob/main/tests/test_0010-start-streamers.py) and [test_0011-generate-classes-from-streamers.py](https://github.com/scikit-hep/uproot5/blob/main/tests/test_0011-generate-classes-from-streamers.py)), the `tojson` method hasn't been used by anything and it hasn't been kept up-to-date with recent developments, such as serialization of additional data types. I've been on the fence about removing it completely: as a public interface, it should either work or not exist, and it would be easier to back off on it than to go forward with it. The reason I haven't done so is because it's a good diagnostic when someone raises a serialization issue.

**Comment by agoose77** [comment]
> I don't think that bare TString objects can be put in TDirectories (even in ROOT, that is). 

Oh, that sounds familiar, actually.

> But perhaps Uproot could use some code to prohibit non-TDirectory-ready objects from being written to TDirectories.

Good to see a silly mistake motivating a new feature ;)

**Comment by lobis** [comment]
After some work I became pretty stuck.

- https://github.com/scikit-hep/uproot5/pull/763/commits/5c90b481ace60ca803bf3e4498dd3f1e1a1f5130 fixed (allegedly) serialization of `TObjString` inside the `TList`.

- **The serialization of `TList` is still wrong and now I am pretty sure the problem is either in `uproot.writing.identify.to_TList` or in `TList::_serialize`**.

- The deserealization (reading) of `TList` appears to work fine. An empty list is serialized / deserialized correctly too.

- I tried to use the simples `TList` I could think of (containing a single `TObject`) to test, and see that the problem is not related to `TObjString`.

A simple root script to generate this basic list:

```c++
auto f = TFile::Open("list.root", "RECREATE");

auto l = new TList();

l->Add(new TObject()); // I also tried setting a non-zero unique ID for the TObject, deserialization still works

l->Write("myList", TObject::kSingleKey); // if TObject::kSingleKey is not specified, each item on the list is written as a key to the root file

f->Close();
```

- This is read correctly by `uproot`.

- **Attempt to recreate this file with uproot**:

```python
import uproot

tobject = uproot.models.TObject.Model_TObject.empty()
tobject._members["@fUniqueID"] = 0
tobject._members["@fBits"] = 0

mylist = uproot.writing.identify.to_TList([tobject])

filename = "list.root"
key = "mylist"

with uproot.recreate(filename) as f:
    f[key] = mylist
```

which will produce an error like:

```
    raise DeserializationError(
uproot.deserialization.DeserializationError: while reading
    TList version 5 as uproot.models.TList.Model_TList (31 bytes)
        (base): <TObject None None at 0x0122b05f10c0>
        fName: ''
        fSize: 1
Base classes for TList: TObject?
Members for TList: fName?, fSize?
expected 31 bytes but cursor moved by 26 bytes (through TList)
in file list.root
in object /mylist;1
```

> Python zip stops when it reaches the end of either of the lists it is given, so filling self._data and not self._options will not get it to write the correct number of items. If that number of items disagrees with the declared self._members["fSize"], the reader will likely crash.

I added a check for this but it's not the problem in this case.

I am not sure how to proceed, unless I am missing something the problem seems a bit too deep for me yet. Any tips are welcome @jpivarski @agoose77 ty.

**Comment by agoose77** [comment]
It might be helpful to write a `TList[TObject()]` from ROOT, and try to round-trip it through uproot. This also fails, but might help to narrow down where the serialisation is differing. (I assume the error lies in how we serialise the `TList` / `TObject`).

**Comment by lobis** [comment]
> fixes the reading.

Thanks @agoose77! this fixed the issue. I implemented your suggestion on the `uproot.writing.identity.to_TList` method instead of in the `TList` model itself because it seemed more appropiate (4da653fee0f28789fbb20abcad3a4541acd6701a).

File: src/uproot/writing/identify.py [hunk 1]
```diff
@@ -761,15 +763,14 @@ def to_TList(data, name=""):
...
-    tobject._members["@fBits"] = 0
 
     tlist = uproot.models.TList.Model_TList.empty()
     tlist._bases.append(tobject)
     tlist._members["fName"] = name
     tlist._data = list(data)
     tlist._members["fSize"] = len(tlist._data)
-    tlist._options = [b""] * len(tlist._data)
+    # see https://github.com/scikit-hep/uproot5/pull/763#issuecomment-1290149135
+    tlist._options = [uproot.serialization.string("")] * len(tlist._data)
```
[the remark this answers is not in the document]
Comment (jpivarski) [reply src/uproot/writing/identify.py:773]: I don't know what `_options` are used for, but they've always been `b""` in all the files/contexts I've seen. Maybe it is a used-to-be-used feature that is today ignored.

File: src/uproot/writing/identify.py [hunk 2]
```diff
@@ -735,12 +735,14 @@ def to_TObjString(string):
...
     """
+    tobject = uproot.models.TObject.Model_TObject.empty()
+
     tobjstring = uproot.models.TObjString.Model_TObjString(str(string))
     tobjstring._deeply_writable = True
     tobjstring._cursor = None
     tobjstring._parent = None
     tobjstring._members = {}
-    tobjstring._bases = (uproot.models.TObject.Model_TObject(),)
+    tobjstring._bases = [tobject]
```
Comment (jpivarski) [inline src/uproot/writing/identify.py:745]: Let's just keep this as a tuple. It's probably a tuple everywhere else.

```suggestion
    tobjstring._bases = (tobject,)
```

It _might_ matter if there's some code that concatenates lists of bases with `+` (`[1] + (2, 3)` is an error), though I don't think that there is. Still, if we just stay consistent with these types, the problem will never come up.

**Review by jpivarski — CHANGES_REQUESTED** [review]
I did a scan through and saw some important fixes introduced by this PR. The tests are good (checking edge cases like empty lists as well as non-empty ones) and I'm not surprised that fixing things like making the options list match the TList size was necessary to get those tests to pass.

The two requested changes are in call-outs, one of them below, and the other is from @agoose77. Maintaining separation-of-concerns is important.

All in all, this looks great, thanks!

File: src/uproot/writing/identify.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lobis) [reply src/uproot/writing/identify.py]: > That's right: only the functions and modules with "serialize" or "write" in their names should be concerned with serialization/writing. identify.py is for generating `Interpretation` objects, to know how to map C++ objects in TTrees into data in arrays.
> 
> The `_options` should continue to be Python objects, `b""`, and only get serialized with the `bytestring` function when the serialization process is already underway (in a `serialize` method).

I implemented the requested change in d6e154308dee74c8a21170578e32af0f20827551 but left the `_options` as python `str`. In your comment you suggest keeping it as python `bytes` instead. It feels wierd to me that in the `serialize` method this is converted back into a `str` only to be serialized in a different way. Specially when other members such as `_members["fName"]` are `str`.

Any reason why the `_options` entries should be `bytes` and not `str` @jpivarski ?

**Comment by lobis** [comment]
> The CI here is failing because we don't upper-bound our Awkward version, and it's picking up the V2 RC candidate. This surprised me, until I realised that this happens when the dependency specifier already includes a pre-release, e.g. `1.9.0rc1`. I recommend we wait until #765 merges, and then rebase this.

Thanks, I was wondering what was wrong.

File: src/uproot/writing/identify.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpivarski) [reply src/uproot/writing/identify.py]: Yeah, I had no idea what they represented, whether they would ever be human-readable or not. That draw `option` in `TObject::SetDrawOption` is not necessarily related to this `option`. I've never seen what TList options do.

Strictly speaking, any `char*` or `std::string` from C++ ought to be a Python `bytes`: those are at the same level of abstraction. In Uproots 1‒3, they were, but users complained about the `bytes` handling, especially as Python 3 became the only Python.

So we have a policy of presenting human-readable/interpretable strings (like `fName` and `fTitle`) as `str` and binary stuff (like `fUUID`) as `bytes`. We use `"utf-8"` with `errors="surrogateescape"` as an encoding convention. It's not necessarily what the user intended—maybe they expect all of their axis labels to be in `"latin-1"`, or maybe `"latin-4"`, but UTF-8 is ASCII for ASCII and `errors="surrogateescape"` allows non-UTF-8 to at least be round-trippable:

```python
>>> b"he\xffllo".decode("utf-8")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 2: invalid start byte
```

but

```python
>>> b"he\xffllo".decode("utf-8", errors="surrogateescape")
'he\udcffllo'
```

and

```python
>>> b"he\xffllo".decode("utf-8", errors="surrogateescape").encode(errors="surrogateescape")
b'he\xffllo'
```

I think there are functions in _util.py for these conversions.

About your real question: wanting to avoid double-conversion. I just noticed that there's no `bytestring` function to complement our

https://github.com/scikit-hep/uproot5/blob/33777b16c654f1fa4c26a960af1e70f2357425d3/src/uproot/serialization.py#L19-L32

You can just add one (also in serialization.py):

```python
def bytestring(data):
    """
    Converts Python bytes into another byte sequence that is ready to be serialized
    into a file.

    If the length of the bytes object is less than 255, it is preceded by a 1-byte
    length; otherwise, it is preceded by ``b'\xff'`` and a 4-byte length.
    """
    length = len(data)
    if length < 255:
        return struct.pack(">B%ds" % length, length, data)
    else:
        return struct.pack(">BI%ds" % length, 255, length, data)
```

File: src/uproot/models/TObject.py [hunk 3]
```diff
@@ -124,5 +127,13 @@ def tojson(self):
             "fBits": self.member("@fBits"),
         }
 
+    @classmethod
+    def empty(cls):
+        self = uproot.model.Model.empty()
+        self.__class__ = cls
```
[the remark this answers is not in the document]
Comment (lobis) [reply src/uproot/models/TObject.py:133]: Yes I totally agree, didn't know how to do it properly, thanks!

File: src/uproot/writing/identify.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpivarski) [reply src/uproot/writing/identify.py]: Indeed it does! I opened up a browser window, went to the appropriate file in GitHub, and didn't see it there—in `main`. I should have thought of also checking _this branch_!

This comment thread can be resolved now because it's no longer mixing serialization with representation:

https://github.com/lobis/uproot5/blob/2d6917490b86a7cb89884d5b08e9bcb7ac0424e1/src/uproot/writing/identify.py#L772

**Review by jpivarski — APPROVED** [review]
This is great! Thanks, @lobis, for this nice work!

**Comment by jpivarski** [comment]
@all-contributors please add @lobis for code
